### PR TITLE
AK: Introduce a new round_to helper, to leverage the HW's rounding capabilities

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
Before re were spending 11-15% of time in `round()` while running [Cookie Clicker](https://orteil.dashnet.org/cookieclicker)
This should remove that

----

Profiling results while showing cookie clicker:
```
i686 + sse2 (master): draw_scaled_bitmap : 30.59% (16.76% self)
    - roundf 11.53%
    - blit 1.98%
    - blit_with_opacity 0.04%
i686 + sse2 (new): draw_scaled_bitmap : 18.01% (12.69% self)
    - blit 4.66%
    - blit_with_opacity 0.01%
```

### CODE REVIEW

**Comment by GMTA** [comment]
Now `Rect::to_rounded()` for integral types no longer rounds but floors the values instead. Why not add new functions for flooring and use them explicitly at the place where you want to improve performance?

**Comment by Hendiadyoin1** [comment]
Ahh forgot that the standard behaviour is supposed to be truncation...

Will look at what the compiler does, and how to get rounding, ideally without calling `llrint`

Will mark this as a draft in the meantime

**Comment by Hendiadyoin1** [comment]
So I've updated it to use an asm helper, to match the intention, and not being forced to follow C99 convention

There are probably more places where this is useful.

File: AK/Math.h [hunk 1]
```diff
@@ -532,6 +532,97 @@ using Hyperbolic::cosh;
...
+        asm(
+            "fistps %0"
+            : "=m"(ret)
+            : "t"(value)
+            : "st");
+    }
+    return ret;
+#else
+    if constexpr (IsSame<P, long double>)
+        return (I)llrintl(value);
```
Comment (GMTA) [inline AK/Math.h:563]: `static_cast<I>` instead of C-style casts?

File: AK/Math.h [hunk 2]
```diff
@@ -532,6 +532,97 @@ using Hyperbolic::cosh;
...
+#endif
+}
+
+#ifdef __SSE__
+template<Integral I>
+ALWAYS_INLINE I round_to(float value)
+{
+#    if ARCH(I386) || ARCH(X86_64)
+#        if ARCH(X86_64)
+    if constexpr (sizeof(I) == sizeof(i64)) {
```
Comment (GMTA) [inline AK/Math.h:577]: Maybe `#elif` instead of duplicating the condition here? Found this a bit confusing to read.

Comment (Hendiadyoin1) [reply AK/Math.h:577]: Well This is not an explicit else,
This part is x64 only, while the following is x86 and x64,
I will reorder the statements, to make it more readeable though

Comment (Hendiadyoin1) [reply AK/Math.h:576]: It should now be a bit clearer why there is this distinction

Comment (GMTA) [reply AK/Math.h:573]: Yeah, this is easier to follow. Thanks!

**Review by linusg — COMMENTED** [review]
Perhaps there's a misunderstanding on my part, but why does this need to be a separate function instead of being part of the built-in functions in LibM? (...beyond the `static_cast` to `I`, which easily could be done at the call site)

**Comment by GMTA** [comment]
@linusg 
> but why does this need to be a separate function instead of being part of the built-in functions in LibM?

My guess is that `AK::round_to` provides a specific, optimized version of rounding similar to how we have other LibC-type functions in the `AK` namespace. In LibC we would probably want to implement rounding similar to how other platforms implement it in order for some compatibility, but we probably can use this new function there as well.

**Comment by Hendiadyoin1** [comment]
I answered this via discord as follows:

> So C99 enforces break tie away from zero, which ensues some more complexity, and in these cases we do not care about the break-tie algorithm, so we could use rint
> the compiler is allowed to replace rint or in these cases lrint (which returns long, which isnt int all the time...), but the compilers do not seem to do this,
> so having a nice looking essentially single instruction way to do this, which is more understandable/shorter than static_cast<int>(lrintf(my_value)), seems appropriate
> 
> Note 1: Also we were doing a normal round and cast before, which first rounded and then tried to truncate again....
> Note 2: This is tailored to our currently specified SSE-capability, while out libm usually lives in x87 land....
> Note 3: 12% less of absolute time  spent in draw_scaled bitmap is a lot

After further reasearch Clang and gcc will replace `lrint` with `cvtss2si`, if we enable `-fno-math-errno`, which would be an alternative+changing a lot of calls to round wiht `[l]rint`

**Comment by Hendiadyoin1** [comment]
I can also adjust our LibM's rint with the new AK helpers, to choose the best impl availible on compile time (important for x86_64 which always has SSE2)

**Pull request merged**